### ISSUE TITLE
9-slice shader fails to link on Mali-400 MP

### ISSUE THREAD

**Issue by willeastcott** [playcanvas/engine#1269]
### Description

This shader pair fails to link on a Mali-400 MP:

Vertex shader:
```
varying vec3 vPositionW;
varying vec2 vUv0;

attribute vec3 vertex_position;
attribute vec3 vertex_normal;
attribute vec4 vertex_tangent;
attribute vec2 vertex_texCoord0;
attribute vec2 vertex_texCoord1;
attribute vec4 vertex_color;
uniform mat4 matrix_viewProjection;
uniform mat4 matrix_model;
uniform mat3 matrix_normal;
vec3 dPositionW;
mat4 dModelMatrix;
mat3 dNormalMatrix;
vec3 dLightPosW;
vec3 dLightDirNormW;
vec3 dNormalW;
#ifndef NINESLICE
#define NINESLICE
uniform vec4 innerOffset;
uniform vec2 outerScale;
uniform vec4 atlasRect;
varying vec2 vTiledUv;
#endif
varying vec2 vMask;
vec2 getUv0() {
    vec2 uv = vertex_position.xz;
    // offset inner vertices inside
    // (original vertices must be in [-1;1] range)
    vec2 positiveUnitOffset = clamp(vertex_position.xz, vec2(0.0), vec2(1.0));
    vec2 negativeUnitOffset = clamp(-vertex_position.xz, vec2(0.0), vec2(1.0));
    uv += (-positiveUnitOffset * innerOffset.xy + negativeUnitOffset * innerOffset.zw) * vertex_texCoord0.xy;
    uv = uv * -0.5 + 0.5;
    uv = uv * atlasRect.zw + atlasRect.xy;
    vMask = vertex_texCoord0.xy;
    return uv;
}
#define NINESLICED
#ifdef PIXELSNAP
    uniform vec4 uScreenSize;
#endif
#ifdef NINESLICED
    #ifndef NINESLICE
    #define NINESLICE
    uniform vec4 innerOffset;
    uniform vec2 outerScale;
    uniform vec4 atlasRect;
    varying vec2 vTiledUv;
    #endif
#endif
mat4 getModelMatrix() {
    #ifdef DYNAMICBATCH
        return getBoneMatrix(vertex_boneIndices);
    #elif defined(SKIN)
        return matrix_model * (getBoneMatrix(vertex_boneIndices.x) * vertex_boneWeights.x +
               getBoneMatrix(vertex_boneIndices.y) * vertex_boneWeights.y +
               getBoneMatrix(vertex_boneIndices.z) * vertex_boneWeights.z +
               getBoneMatrix(vertex_boneIndices.w) * vertex_boneWeights.w);
    #elif defined(INSTANCING)
        return mat4(instance_line1, instance_line2, instance_line3, instance_line4);
    #else
        return matrix_model;
    #endif
}
vec4 getPosition() {
    dModelMatrix = getModelMatrix();
    vec3 localPos = vertex_position;
    #ifdef NINESLICED
        // outer and inner vertices are at the same position, scale both
        localPos.xz *= outerScale;
        // offset inner vertices inside
        // (original vertices must be in [-1;1] range)
        vec2 positiveUnitOffset = clamp(vertex_position.xz, vec2(0.0), vec2(1.0));
        vec2 negativeUnitOffset = clamp(-vertex_position.xz, vec2(0.0), vec2(1.0));
        localPos.xz += (-positiveUnitOffset * innerOffset.xy + negativeUnitOffset * innerOffset.zw) * vertex_texCoord0.xy;
        vTiledUv = (localPos.xz - outerScale + innerOffset.xy) * -0.5 + 1.0; // uv = local pos - inner corner
        localPos.xz *= -0.5; // move from -1;1 to -0.5;0.5
        localPos = localPos.xzy;
    #endif
    vec4 posW = dModelMatrix * vec4(localPos, 1.0);
    #ifdef SCREENSPACE
        posW.zw = vec2(0.0, 1.0);
    #endif
    dPositionW = posW.xyz;
    vec4 screenPos;
    #ifdef UV1LAYOUT
        screenPos = vec4(vertex_texCoord1.xy * 2.0 - 1.0, 0.5, 1);
    #else
        #ifdef SCREENSPACE
            screenPos = posW;
        #else
            screenPos = matrix_viewProjection * posW;
        #endif
        #ifdef PIXELSNAP
            // snap vertex to a pixel boundary
            screenPos.xy = (screenPos.xy * 0.5) + 0.5;
            screenPos.xy *= uScreenSize.xy;
            screenPos.xy = floor(screenPos.xy);
            screenPos.xy *= uScreenSize.zw;
            screenPos.xy = (screenPos.xy * 2.0) - 1.0;
        #endif
    #endif
    return screenPos;
}
vec3 getWorldPosition() {
    return dPositionW;
}


void main(void) {
    gl_Position = getPosition();
   vPositionW    = getWorldPosition();
   vec2 uv0 = getUv0();
   vUv0 = uv0;
}
```

Fragment shader:
```
#extension GL_OES_standard_derivatives : enable

precision mediump float;
varying vec3 vPositionW;
varying vec2 vUv0;

uniform vec3 view_position;
uniform vec3 light_globalAmbient;
float square(float x) {
    return x*x;
}
float saturate(float x) {
    return clamp(x, 0.0, 1.0);
}
vec3 saturate(vec3 x) {
    return clamp(x, vec3(0.0), vec3(1.0));
}
#define NINESLICETILED
varying vec2 vMask;
varying vec2 vTiledUv;
uniform vec4 innerOffset;
uniform vec2 outerScale;
uniform vec4 atlasRect;
vec2 nineSlicedUv;vec4 dReflection;
vec3 dAlbedo;
vec3 dDiffuseLight;
vec3 dSpecularLight;
vec3 dSpecularity;
float dAlpha;

vec4 texture2DSRGB(sampler2D tex, vec2 uv) {
    return texture2D(tex, uv);
}
vec4 texture2DSRGB(sampler2D tex, vec2 uv, float bias) {
    return texture2D(tex, uv, bias);
}
vec4 textureCubeSRGB(samplerCube tex, vec3 uvw) {
    return textureCube(tex, uvw);
}
vec3 gammaCorrectOutput(vec3 color) {
    return color;
}
vec3 gammaCorrectInput(vec3 color) {
    return color;
}
float gammaCorrectInput(float color) {
    return color;
}
vec4 gammaCorrectInput(vec4 color) {
    return color;
}
vec3 toneMap(vec3 color) {
    return color;
}
vec3 addFog(vec3 color) {
    return color;
}

#undef MAPFLOAT

#undef MAPCOLOR
 #define MAPCOLOR

#undef MAPVERTEX

#undef MAPTEXTURE
#ifdef MAPCOLOR
uniform vec3 material_diffuse;
#endif
#ifdef MAPTEXTURE
uniform sampler2D texture_diffuseMap;
#endif
void getAlbedo() {
    dAlbedo = vec3(1.0);
    #ifdef MAPCOLOR
        dAlbedo *= material_diffuse.rgb;
    #endif
    #ifdef MAPTEXTURE
        dAlbedo *= texture2DSRGB(texture_diffuseMap, UV).CH;
    #endif
    #ifdef MAPVERTEX
        dAlbedo *= gammaCorrectInput(saturate(vVertexColor.VC));
    #endif
}

#undef MAPFLOAT
 #define MAPFLOAT

#undef MAPCOLOR

#undef MAPVERTEX

#undef MAPTEXTURE
 #define MAPTEXTURE
#ifdef MAPFLOAT
uniform float material_opacity;
#endif
#ifdef MAPTEXTURE
uniform sampler2D texture_opacityMap;
#endif
void getOpacity() {
    dAlpha = 1.0;
    #ifdef MAPFLOAT
        dAlpha *= material_opacity;
    #endif
    #ifdef MAPTEXTURE
        dAlpha *= texture2D(texture_opacityMap, nineSlicedUv, -1000.0).a;
    #endif
    #ifdef MAPVERTEX
        dAlpha *= saturate(vVertexColor.VC);
    #endif
}

#undef MAPFLOAT

#undef MAPCOLOR
 #define MAPCOLOR

#undef MAPVERTEX

#undef MAPTEXTURE
 #define MAPTEXTURE
#ifdef MAPCOLOR
uniform vec3 material_emissive;
#endif
#ifdef MAPFLOAT
uniform float material_emissiveIntensity;
#endif
#ifdef MAPTEXTURE
uniform sampler2D texture_emissiveMap;
#endif
vec3 getEmission() {
    vec3 emission = vec3(1.0);
    #ifdef MAPFLOAT
        emission *= material_emissiveIntensity;
    #endif
    #ifdef MAPCOLOR
        emission *= material_emissive;
    #endif
    #ifdef MAPTEXTURE
        emission *= texture2DSRGB(texture_emissiveMap, nineSlicedUv, -1000.0).rgb;
    #endif
    #ifdef MAPVERTEX
        emission *= gammaCorrectInput(saturate(vVertexColor.VC));
    #endif
    return emission;
}
vec3 combineColor() {
    return dAlbedo * dDiffuseLight;
}

void addAmbient() {
    dDiffuseLight += light_globalAmbient;
}

void main(void) {
    dDiffuseLight = vec3(0);
    dSpecularLight = vec3(0);
    dReflection = vec4(0);
    dSpecularity = vec3(0);
vec2 tileMask = step(vMask, vec2(0.99999));
vec2 clampedUv = mix(innerOffset.xy*0.5, vec2(1.0) - innerOffset.zw*0.5, fract(vTiledUv));
clampedUv = clampedUv * atlasRect.zw + atlasRect.xy;
nineSlicedUv = vUv0 * tileMask + clampedUv * (vec2(1.0) - tileMask);   getOpacity();
   getAlbedo();
   addAmbient();

   gl_FragColor.rgb = combineColor();
   gl_FragColor.rgb += getEmission();
   gl_FragColor.rgb = addFog(gl_FragColor.rgb);
   #ifndef HDR
    gl_FragColor.rgb = toneMap(gl_FragColor.rgb);
    gl_FragColor.rgb = gammaCorrectOutput(gl_FragColor.rgb);
   #endif
gl_FragColor.rgb *= dAlpha;
gl_FragColor.a = dAlpha;

}
```

Note that when the engine runs on the Mali-400 MP, it reports:

```
WARNING: highp not supported, using mediump
```

This shader is unusual because it declares a uniform in _both_ the vertex and fragment shader. We explicitly specific a default precision in the fragment shader, but not in the vertex shader. Despite being a mediump GPU, I'm assuming the vertex shader is being compiled with a default precision of highp. So the atlasRect uniform appears in both vertex and fragment shaders with a different precision.

Therefore the fix is probably to also set a default precision in our vertex shaders.

**Comment by willeastcott** [comment]
Hey @guycalledfrank - before I have a stab at fixing this, do you have any thoughts?

**Comment by guycalledfrank** [comment]
Any shader compilation errors, or just this warning?

**Comment by willeastcott** [comment]
Error is:

```
launch.js:7522 ERROR:   Failed to link shader program. Error: Uniforms with the same name but different type/precision: atlasRect
```

**Comment by guycalledfrank** [comment]
The quickest way to fix it is to just rename the uniform in the fragment shader. That would work for mixed precision scenarios, but perhaps will add a little overhead to other platforms.
Option 2: just force declare it as "uniform mediump vec4" everywhere (are there any devices with lowp?!). I suspect medium precision should be fine for atlas UVs.

**Comment by guycalledfrank** [comment]
That can give some precision problems, e.g. jittery vertices. We'd better run VS in the best precision we can..

**Comment by willeastcott** [comment]
OK, good point! I didn't consider that. I've added some comments to this PR to fix this:

https://github.com/playcanvas/engine/pull/1251/files

Make it so, @daredevildave. 😄

**Comment by daredevildave** [comment]
Fixed in #1251
